src/internet/spotifywebapi/spotifywebapiservice.cpp

```diff
@@ -0,0 +1,189 @@
+/* This file is part of Clementine.
+   Copyright 2021, Kenman Tsang <kentsangkm@pm.me>
+
+   Clementine is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Clementine is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "spotifywebapiservice.h"
+
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QtDebug>
+#include <utility>
+
+#include "3rdparty/qtiocompressor/qtiocompressor.h"
+#include "core/application.h"
+#include "core/network.h"
+#include "globalsearch/globalsearch.h"
+#include "globalsearch/spotifywebapisearchprovider.h"
+#include "ui/iconloader.h"
+
+static constexpr auto kServiceName = "SpotifyWebApi";
+
+namespace {
+template <typename T>
+inline const T& Get(const T& obj) {
+  return obj;
+}
+
+template <typename Name, typename... Args>
+inline QJsonValue Get(const QJsonValue& obj, Name&& name, Args&&... args) {
+  Q_ASSERT(obj.isObject());
+  return Get(obj.toObject()[std::forward<Name>(name)],
+             std::forward<Args>(args)...);
+}
+
+template <typename Name, typename... Args>
+inline QJsonValue Get(const QJsonDocument& obj, Name&& name, Args&&... args) {
+  return Get(obj.object()[std::forward<Name>(name)],
+             std::forward<Args>(args)...);
+}
+}  // namespace
+
+SpotifyWebApiService::SpotifyWebApiService(Application* app,
+                                           InternetModel* parent)
+    : InternetService(kServiceName, app, parent, parent),
+      network_(new NetworkAccessManager{this}) {
+  app_->global_search()->AddProvider(
+      new SpotifyWebApiSearchProvider(app_, this));
+}
+
+SpotifyWebApiService::~SpotifyWebApiService() {}
+
+QStandardItem* SpotifyWebApiService::CreateRootItem() {
+  root_ = new QStandardItem(IconLoader::Load("spotify", IconLoader::Provider),
+                            kServiceName);
+  return root_;
+}
+
+void SpotifyWebApiService::LazyPopulate(QStandardItem* item) {}
+
+void SpotifyWebApiService::Search(int searchId, QString queryStr) {
+  if (std::chrono::system_clock::now().time_since_epoch() >= tokenExpiration_) {
```

Comment:
`QTime` might be easier to work with.

Reply:
But I cannot see any QTime or QDateTime ctor with milliseconds or any similar concept.